presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -85,441 +94,708 @@
+            HiveColumnHandle column,
+            Type type,
+            List<HivePartition> partitions,
+            Map<String, PartitionStatistics> statistics,
+            double rowsPerPartition)
     {
-        if (rowCount.isUnknown()) {
+        if (!isVarcharType(type) && !isCharType(type)) {
             return Estimate.unknown();
         }
-        if (rowCount.getValue() == 0.0) {
-            return Estimate.zero();
+        double dataSize = 0;
+        for (HivePartition partition : partitions) {
+            int length = getSize(partition.getKeys().get(column));
+            double rowCount = getPartitionRowCount(partition.getPartitionId(), statistics).orElse(rowsPerPartition);
+            dataSize += length * rowCount;
         }
+        return Estimate.of(dataSize);
+    }
 
-        Estimate totalNullsCount = summarizePartitionStatistics(
-                statisticsSample.values(),
-                column,
-                columnStatistics -> {
-                    if (!columnStatistics.getNullsCount().isPresent()) {
-                        return OptionalDouble.empty();
-                    }
-                    return OptionalDouble.of(columnStatistics.getNullsCount().getAsLong());
-                },
-                nullsCountStream -> {
-                    double nullsCount = 0;
-                    long partitionsWithStatisticsCount = 0;
-                    for (PrimitiveIterator.OfDouble nullsCountIterator = nullsCountStream.iterator(); nullsCountIterator.hasNext(); ) {
-                        nullsCount += nullsCountIterator.nextDouble();
-                        partitionsWithStatisticsCount++;
-                    }
-
-                    if (partitionsWithStatisticsCount == 0) {
-                        return OptionalDouble.empty();
-                    }
-                    return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
-                });
+    private static int getSize(NullableValue nullableValue)
+    {
+        if (nullableValue.isNull()) {
+            return 0;
+        }
+        Object value = nullableValue.getValue();
+        checkArgument(value instanceof Slice, "value is expected to be of Slice type");
+        return ((Slice) value).length();
+    }
 
-        if (totalNullsCount.isUnknown()) {
-            return Estimate.unknown();
+    private static OptionalDouble getPartitionRowCount(String partitionName, Map<String, PartitionStatistics> statistics)
+    {
+        PartitionStatistics partitionStatistics = statistics.get(partitionName);
+        if (partitionStatistics == null) {
+            return OptionalDouble.empty();
         }
-        return Estimate.of(totalNullsCount.getValue() / rowCount.getValue());
+        OptionalLong rowCount = partitionStatistics.getBasicStatistics().getRowCount();
+        if (rowCount.isPresent()) {
+            verify(rowCount.getAsLong() >= 0, "rowCount must be greater than or equal to zero");
```

Comment:
Why so? There could be empty partitions.

Reply:
Than the row count will be `0`.